File: src/commands/key.ts
import { resolveProjectRoot } from '../utils/paths';
import { success, fail, cleanOutput } from '../utils/output';
import { readSession, isPtyAlive } from '../pty/pty-manager';
import { enqueueInput } from '../queue/queue-engine';
import { getBufferSize, readBufferSince } from '../buffer/buffer-manager';
import { restoreSession } from '../runtime/restore';
import { logEvent } from '../ledger/ledger';
import {
  sessionNotFoundError,
  sessionNotRunningError,
} from '../utils/validate';

// ─── Named key → escape sequence map ────────────────────────────────────────

const KEY_MAP: Record<string, string> = {
  up: '\x1b[A',
  down: '\x1b[B',
  right: '\x1b[C',
  left: '\x1b[D',
  enter: '\r',
  return: '\r',
  tab: '\t',
  escape: '\x1b',
  esc: '\x1b',
  space: ' ',
  backspace: '\x7f',
  delete: '\x1b[3~',
  home: '\x1b[H',
  end: '\x1b[F',
  pageup: '\x1b[5~',
  pagedown: '\x1b[6~',
  // Ctrl combos
  'ctrl-c': '\x03',
  'ctrl-d': '\x04',
  'ctrl-z': '\x1a',
  'ctrl-l': '\x0c',
  'ctrl-a': '\x01',
  'ctrl-e': '\x05',
  // "y" and "n" shortcuts for confirm prompts
  y: 'y',
  n: 'n',
};

/**
 * Resolve a key name to its byte sequence.
 * Supports names from KEY_MAP, or raw text like "abc".
 */
function resolveKey(name: string): string | null {
  const lower = name.toLowerCase();
  return KEY_MAP[lower] ?? null;
}

export async function keyCommand(
  terminalId: string,
  keys: string[]
): Promise<void> {
  const projectRoot = resolveProjectRoot();

  // ── Validate keys ──────────────────────────────────────────────────────
  const resolved: string[] = [];
  const unknown: string[] = [];

  for (const k of keys) {
    const seq = resolveKey(k);
    if (seq !== null) {
      resolved.push(seq);
    } else {
      unknown.push(k);
    }
  }

  if (unknown.length > 0) {
    fail({
      error: `Unknown key name(s): ${unknown.join(', ')}`,
      hints: {
        available_keys: Object.keys(KEY_MAP).join(', '),
        example: 'clrun key <id> down down enter',
        note: 'Keys are case-insensitive. Use "clrun input" for text input.',
      },
    });
  }

  // ── Verify session exists ──────────────────────────────────────────────
  let session = readSession(terminalId, projectRoot);
  if (!session) {
    fail(sessionNotFoundError(terminalId));
  }

  // ── Transparent restore for suspended sessions ─────────────────────────
  if (session!.status === 'suspended') {
    // Queue the key sequence as a raw string (worker will handle \r)
    // But keys are raw escape sequences — we need a special queue entry
    // that the worker sends without appending \r.
    // For now, enqueue each key individually so the worker sends each + \r,
    // but we actually need the raw bytes. Let's write directly via SIGUSR1
    // after restore.
    await restoreSession(terminalId, projectRoot);
    session = readSession(terminalId, projectRoot);
    await new Promise((r) => setTimeout(r, 300));
  }

  // ── Check session is running ───────────────────────────────────────────
  if (session!.status !== 'running') {
    fail(sessionNotRunningError(terminalId, session!.status));
  }

  if (!isPtyAlive(session!.worker_pid)) {
    fail({
      error: `Session worker is not alive (PID: ${session!.worker_pid})`,
      hints: {
        check_status: 'clrun status',
        start_new: 'clrun <command>',
      },
    });
  }

  // ── Send key sequences via the queue ───────────────────────────────────
  // Concatenate all resolved keys into one raw string and enqueue it.
  // The worker will append \r, but for raw key sequences we need to
  // mark it so the worker sends it verbatim.
  const rawSequence = resolved.join('');

  // We use a special prefix that the worker recognizes to send raw bytes
  // without appending \r.
  const RAW_PREFIX = '\x00RAW\x00';
  enqueueInput(terminalId, RAW_PREFIX + rawSequence, 999, projectRoot);

  // Signal the worker to process immediately
  try { process.kill(session!.worker_pid, 'SIGUSR1'); } catch {}

  logEvent('key.sent', projectRoot, terminalId, {
    keys: keys,
    sequence_length: rawSequence.length,
  });

  // ── Wait for output ────────────────────────────────────────────────────
  const bufferBefore = getBufferSize(terminalId, projectRoot);
  await new Promise((r) => setTimeout(r, 400));

  const newLines = readBufferSince(terminalId, bufferBefore, projectRoot);
  const output = cleanOutput(newLines);

  success({
    terminal_id: terminalId,
    keys_sent: keys,
    ...(output && { output }),
    hints: {
      send_more_keys: `clrun key ${terminalId} <key> [<key>...]`,
      send_text: `clrun ${terminalId} '<text>'`,
      view_output: `clrun tail ${terminalId} --lines 50`,
      available_keys: 'up, down, left, right, enter, tab, escape, space, backspace, ctrl-c, ctrl-d',
    },
  });
}

File: python/clrun/commands/key.py
"""The `clrun key` command — send named keystrokes to a running session."""

from __future__ import annotations

import os
import signal
import time
from typing import List

from clrun.utils.paths import resolve_project_root
from clrun.utils.output import success, fail, clean_output
from clrun.pty.pty_manager import read_session, is_pty_alive
from clrun.queue.queue_engine import enqueue_input
from clrun.buffer.buffer_manager import get_buffer_size, read_buffer_since
from clrun.runtime.restore import restore_session
from clrun.ledger.ledger import log_event
from clrun.utils.validate import session_not_found_error, session_not_running_error

KEY_MAP = {
    "up": "\x1b[A",
    "down": "\x1b[B",
    "right": "\x1b[C",
    "left": "\x1b[D",
    "enter": "\r",
    "return": "\r",
    "tab": "\t",
    "escape": "\x1b",
    "esc": "\x1b",
    "space": " ",
    "backspace": "\x7f",
    "delete": "\x1b[3~",
    "home": "\x1b[H",
    "end": "\x1b[F",
    "pageup": "\x1b[5~",
    "pagedown": "\x1b[6~",
    "ctrl-c": "\x03",
    "ctrl-d": "\x04",
    "ctrl-z": "\x1a",
    "ctrl-l": "\x0c",
    "ctrl-a": "\x01",
    "ctrl-e": "\x05",
    "y": "y",
    "n": "n",
}

RAW_PREFIX = "\x00RAW\x00"


def _resolve_key(name: str) -> str | None:
    return KEY_MAP.get(name.lower())


def key_command(terminal_id: str, keys: List[str]) -> None:
    project_root = resolve_project_root()

    resolved: List[str] = []
    unknown: List[str] = []

    for k in keys:
        seq = _resolve_key(k)
        if seq is not None:
            resolved.append(seq)
        else:
            unknown.append(k)

    if unknown:
        fail({
            "error": f"Unknown key name(s): {', '.join(unknown)}",
            "hints": {
                "available_keys": ", ".join(KEY_MAP.keys()),
                "example": "clrun key <id> down down enter",
                "note": 'Keys are case-insensitive. Use "clrun input" for text input.',
            },
        })
        return

    session = read_session(terminal_id, project_root)
    if not session:
        fail(session_not_found_error(terminal_id))
        return

    if session.status == "suspended":
        restore_session(terminal_id, project_root)
        session = read_session(terminal_id, project_root)
        time.sleep(0.3)

    if session and session.status != "running":
        fail(session_not_running_error(terminal_id, session.status))
        return

    if not is_pty_alive(session.worker_pid):
        fail({
            "error": f"Session worker is not alive (PID: {session.worker_pid})",
            "hints": {
                "check_status": "clrun status",
                "start_new": "clrun <command>",
            },
        })
        return

    raw_sequence = "".join(resolved)
    enqueue_input(terminal_id, RAW_PREFIX + raw_sequence, 999, project_root)

    try:
        os.kill(session.worker_pid, signal.SIGUSR1)
    except OSError:
        pass

    log_event("key.sent", project_root, terminal_id, {
        "keys": keys,
        "sequence_length": len(raw_sequence),
    })

    buffer_before = get_buffer_size(terminal_id, project_root)
    time.sleep(0.4)

    new_lines = read_buffer_since(terminal_id, buffer_before, project_root)
    output = clean_output(new_lines)

    success({
        "terminal_id": terminal_id,
        "keys_sent": keys,
        **({"output": output} if output else {}),
        "hints": {
            "send_more_keys": f"clrun key {terminal_id} <key> [<key>...]",
            "send_text": f"clrun {terminal_id} '<text>'",
            "view_output": f"clrun tail {terminal_id} --lines 50",
            "available_keys": "up, down, left, right, enter, tab, escape, space, backspace, ctrl-c, ctrl-d",
        },
    })
